include/nlohmann/json.hpp

```diff
@@ -3555,15 +3799,126 @@ class basic_json
         return value(ptr, string_t(default_value));
     }
 
+    /*!
+    @brief return the binary subtype
+
+    Returns the numerical subtype of the JSON value, if the JSON value is of
+    type "binary", and it has a subtype.  If it does not have a subtype this
+    function will return 0 as a sentinal value.  However, 0 is a valid subtype,
+    so this function ought not be used to determine whether or not a binary
+    value has a subtype.  You should use `has_subtype` for that.
+
+    @return the numerical subtype of the binary JSON value
+
+    @complexity Constant.
+
+    @exceptionsafety No-throw guarantee: this member function never throws
+    exceptions.
+
+    @sa @ref has_subtype() -- returns whether or not the binary value has a
+    subtype
+
+    @since version 3.7.0
+    */
+    uint8_t subtype() const noexcept
+    {
+        if (is_binary() && m_value.binary->_has_subtype)
+        {
+            return m_value.binary->_subtype;
+        }
+
+        // Should this throw instead?
+        return {};
+    }
+
+    /*!
+    @brief return a mutable reference to the binary subtype
+
+    Returns a mutable reference to the numerical subtype of the JSON value, if
+    the JSON value is of type "binary".  Otherwise it throws a type_error.  If
+    it does not have a valid subtype, this function will force the subtype valid
+    as a side-effect.  Use the non-const version of `has_subtype` to turn it
+    back off if needed.
+
+    @return a mutable reference to the numerical subtype of the binary JSON
+    value
+
+    @complexity Constant.
+
+    @since version 3.7.0
+    */
+
+    uint8_t& subtype()
+    {
+        if (is_binary())
+        {
+            m_value.binary->_has_subtype = true;
+            return m_value.binary->_subtype;
+        }
+
+        JSON_THROW(type_error::create(318, "cannot call subtype() with JSON object of type " + std::string(type_name())));
+    }
+
+    /*!
+    @brief return whether or not the binary subtype has a value
+
+    Returns whether or not the binary subtype has a value.
+
+    @return whether or not the binary subtype has a value.
+
+    @complexity Constant.
+
+    @exceptionsafety No-throw guarantee: this member function never throws
+    exceptions.
+
+    @since version 3.7.0
+    */
+    bool has_subtype() const noexcept
+    {
+        if (is_binary())
+        {
+            return m_value.binary->_has_subtype;
+        }
+
+        // Should this throw instead?
+        return {};
+    }
+
+    /*!
+    @brief return a mutable reference to the boolean that determines whether or
+    not binary subtype has a value.
+
+    Returns a mutable reference to the subtype indicating bool of the JSON
+    value, if the JSON value is of type "binary".  Otherwise it throws a
+    type_error.
+
+    @return a mutable reference to the subtype indicating bool of the binary
+    JSON value
+
+    @complexity Constant.
+
+    @since version 3.7.0
+    */
+
+    bool& has_subtype()
```

Comment:
It's useful for changing whether or not this binary data has a subtype.  For instance, if I read in a BSON (which always has a subtype), and I want to serialize to msgpack's bin (rather than the ext type), I would have to set the `has_subtype` value to false.

Reply:
Thanks for the hint, now I understand.

Then I think it would be nicer to have an API that abstracts away from the simulated optional. As I understand, we would need the following functions:

```cpp
bool has_subtype() const noexcept;
void set_subtype(std::uint8_t subtype) noexcept;
void clear_subtype() noexcept;
std::size_t get_subtype() const noexcept;
```

The latter function could then return `std::size_t(-1)` as value for "no subtype" (just like `std::string::npos`).

What do you think?

Reply:
Sounds good to me.  I'll implement that.

Reply:
Implemented in 38f8e86